ambry-store/src/main/java/com/github/ambry/store/StorageManager.java

```diff
@@ -161,6 +167,35 @@ public StorageManager(StoreConfig storeConfig, DiskManagerConfig diskManagerConf
     }
   }
 
+  /**
+   * Checks whether the replicas are placed on the correct disks. If not, reshuffle the disks, write
+   * the new state to Helix and exit. We assume that this ambry-server instance will then be restarted with the
+   * new disk order that we saved to Helix.
+   * @param diskToReplicas A map of disks to the replicas on those disks.
+   */
+  // Tommy: Make this protected so we can test it in StorageManagerTest?!?!!?!!?
+  protected void reshuffleDisksAndMaybeExit(Map<DiskId, List<ReplicaId>> diskToReplicas) {
+    //System.exit(0);
+    PartitionFinder partitionFinder = new PartitionFinder();
+    Map<DiskId, Set<String>> disksToPartitions = new HashMap<>();
+    for (DiskId currentDisk : diskToReplicas.keySet()) {
+      disksToPartitions.put(currentDisk, partitionFinder.findPartitionsOnDisk(currentDisk));
+    }
+
+    ReplicaPlacementValidator placementValidator = new ReplicaPlacementValidator(diskToReplicas, disksToPartitions);
+    Map<DiskId, DiskId> disksToReshuffle = placementValidator.reshuffleDisks();
+    if (!disksToReshuffle.isEmpty()) {
+      logger.info("Disks need to be reshuffled: {}", disksToReshuffle);
+      if(primaryClusterParticipant.setDisksOrder(disksToReshuffle)) {
+        logger.info(this.getClass().getSimpleName() + " - successfully reshuffled disks. Now terminating"
```

Comment:
we don't need to add class name to the logger, logger would print it by default.

Reply:
Ah good to know. Thanks for pointing that out. This has been updated.